GenericsTests/VennDiagramTests.cs

```diff
@@ -0,0 +1,176 @@
+    }
+
+    [Fact]
+    public void Add_NullCircle_ThrowsNullException()
+    {
+        Assert.Throws<ArgumentNullException>(() => { new VennDiagram<string>("TestData",3).Add(null!); });
+    }
+    [Fact]
+    public void Add_ValidCircles_AddsSuccessfully()
+    {
+        // Arrange
+        Circle<string> circleOne = new("Circle4");
+        VennDiagram<string> vennDiagram = new("ValidData");
+        vennDiagram.Add(circleOne);
+
+        // Act
+        Circle<string> circleTwo = new("Circle5");
+        vennDiagram.Add(circleTwo);
+
+        // Assert
+        Assert.Contains(circleTwo, vennDiagram.Circles);
+    }
+
+    [Fact]
+    public void Add_DuplicateNameCircle_ThrowsArgumentException()
+    {
+        Assert.Throws<ArgumentException>(() => VennDiagram.Add(new Circle<string>("Superheroes")));
+    }
+
+    [Fact]
+    public void Add_OverLimitCircle_ThrowsInvalidOperationException()
+    {
+        Assert.Throws<InvalidOperationException>(() => VennDiagram.Add(new Circle<string>("Cops")));
+    }
+
+    [Fact]
+    public void ToString_NullCircle_PrintsCorrectly()
+    {
+
+        VennDiagram<string> vennDiagram = new("TestData");
+
+        // Assert
+        Assert.Equal($"TestData: {{}}", vennDiagram.ToString());
+    }
+
+    [Fact]
+    public void ToString_ValidCircles_PrintsCorrectly()
+    {
+        Circle<string> circle1 = new("Test1", ["Hi", "Hello"]);
+        Circle<string> circle2 = new("Test2", ["Bojour", "Ahola"]);
+
+        List<Circle<string>> circles = [circle1, circle2];
+        VennDiagram<string> vennDiagram = new("TestData", 2);
+        vennDiagram.Add(circle1);
+        vennDiagram.Add(circle2);
+
+        // Assert
+        Assert.Equal($"{circle1}, {circle2}", vennDiagram.ToString());
+    }
+
+    [Fact]
+    public void Intersection_ValidCirclesNames_ReturnsCircle()
+    {
+        // Assert
+        Circle<string> objectsCircle = VennDiagram.Intersection("New York Superheros", "New Yorkers", "Superheroes");
+        Assert.Equal(3,objectsCircle.Elements.Count);
+        Assert.Contains("Peter", objectsCircle.Elements);
+        Assert.Contains("Michaelangelo", objectsCircle.Elements);
+        Assert.Contains("Tony", objectsCircle.Elements);
+
+    }
+
+    [Fact]
+    public void Intersection_NullName_ThrowsArgumentNullException()
+    {
+
+        Exception execption = Assert.Throws<ArgumentNullException>(() => VennDiagram.Intersection(null!, "New Yorkers", "Superheroes"));
+    }
+
+    [Fact]
+    public void Intersection_NullFirstCircleName_ThrowsArgumentNullException()
+    {
+        Exception execption = Assert.Throws<ArgumentNullException>(() => VennDiagram.Intersection("Null Wizards", null!, "Wizards"));
+    }
+
+    [Fact]
+    public void Intersection_NullSecondCircleName_ThrowsArgumentNullException()
+    {
+        Exception execption = Assert.Throws<ArgumentNullException>(() => VennDiagram.Intersection("Null Wizards", "Wizards", null!)) ;
+    }

```

Comment:
```suggestion
    [Fact]
    public void Intersection_NullName_ThrowsArgumentNullException()
    {

        Assert.Throws<ArgumentNullException>(() => VennDiagram.Intersection(null!, "New Yorkers", "Superheroes"));
    }

    [Fact]
    public void Intersection_NullFirstCircleName_ThrowsArgumentNullException()
    {
        Assert.Throws<ArgumentNullException>(() => VennDiagram.Intersection("Null Wizards", null!, "Wizards"));
    }

    [Fact]
    public void Intersection_NullSecondCircleName_ThrowsArgumentNullException()
    {
        Assert.Throws<ArgumentNullException>(() => VennDiagram.Intersection("Null Wizards", "Wizards", null!)) ;
    }

```
These tests don't do anything I believe. You should either Assert that Exception E is type of ArgumentNullException. This can be done using Assert.isType<ArgumentNullException>(exception). Or you just remove setting them to exception. I'd prefer dong the latter since that's how we did it in class before and on previous assignments.

Reply:
I think asserting that the exception is actually the type you expect is valid. Aka, like for an argumentNullException, check that param passed into the exception is what you expect to be thrown for this